optional slash

akotlar opened the pull request.

cseed commented on monitoring/router.nginx.conf:
```diff
@@ -1,17 +1,17 @@
     server_name _;
 
-    location /monitoring/grafana/ {
+    location ~ ^/monitoring/grafana(/)?$ {
```
You can just remove the trailing slash.

akotlar replied:
but then it would match http://domain2.com/monitoring/grafanaddd

akotlar replied:
Specifically without a = nginx will perform a prefix match

cseed replied:
Specifically, location blocks without a modifier match the URL prefix.

akotlar replied:
I think I don't understand how this is different from what I wrote.

cseed replied:
This won't allow a path after the prefix, e.g. /monitoring/grafana/foo/bar.  I think what you'd want it `^/monitoring/grafana($|/)`.  However, the point of setting Host in the gateway to include the target service is exactly so we can use it to dispatch in nginx.  So I'd change this to use separate server blocks like we do in other routers.

tpoterba commented:
bump

cseed commented:
This should use distinct server blocks rather than routing based on the URL.

akotlar commented:
Ill look at it later today

akotlar commented:
Got blocked on dev deploy not working (500). Will try again in the morning.

cseed commented:
So error propagation from CI back to hailctl isn't great right now.  (Something worth fixing!)  If it an error in what you're trying to deploy (e.g. branch not found, syntax error in build.yaml, etc.) you can find it in the CI log.

FYI, you can't dev deploy monitoring.  It's part of the infrastructure.

akotlar commented:
> So error propagation from CI back to hailctl isn't great right now. (Something worth fixing!) If it an error in what you're trying to deploy (e.g. branch not found, syntax error in build.yaml, etc.) you can find it in the CI log.

I should fix it!

cseed commented:
> I should fix it!

That'd be awesome!

Yes, the steps are steps from build.yaml.  Some background on the current setup, there are two classes of deployed stuff: normal, and "infrastructure".  Infrastructure is stuff that can't be virtualized (or we've chosen not to) within the k8s cluster and can't be tested with the current CI setup.  An example is gateway, that binds the live IP address for hail.is.  Infrastructure will need a "meta" testing setup that will spin up another k8s cluster or testing in a separate staging cluster, where taking down the test cluster isn't an issue.  Obviously, we haven't built this yet.

akotlar commented:
> > I should fix it!
> 
> That'd be awesome!
> 
> Yes, the steps are steps from build.yaml. Some background on the current setup, there are two classes of deployed stuff: normal, and "infrastructure". Infrastructure is stuff that can't be virtualized (or we've chosen not to) within the k8s cluster and can't be tested with the current CI setup. An example is gateway, that binds the live IP address for hail.is. Infrastructure will need a "meta" testing setup that will spin up another k8s cluster or testing in a separate staging cluster, where taking down the test cluster isn't an issue. Obviously, we haven't built this yet.

Thank you for the background. Makes sense

akotlar commented:
Added server blocks. @cseed. Added the proxy forwarding headers for consistency (and may provide more information in logs), although they're not strictly necessary.

Prometheus doesn't seem to be working, but behavior is identical without move to server blocks (namely it redirects to a default Nginx page on internal.hail.is ; same behavior with and without this change).

Behavior of redirecting to "service".internal if missing slash still occurs; this seems to occur without hitting the namespace monitoring router (meaning `k logs router-868b794f58-r49hr -n monitoring` shows nothing). So this appears to be happening upstream.

Had surprising amount of trouble /monitoring from the routes, even with corresponding changes in monitoring.yaml, and trying to rewrite in a /monitoring block (meaning tried location / and location /monitoring/*, both with and without rewrite rule `rewrite /monitoring/grafana/ /` or similar with a capture group). Something I don't quite understand, insight appreciated because I would prefer not to spend more time experimenting with this.

Also, would it be reaonsalbe to not propagate the /namespace/service to internal routes (so rewrite before sending)? It seems like internal server blocks receive the full url, which means that they would need to handle those subpaths when used internally, but not when used normally (for instance I'm not sure how notebook deployed to a namespace gets away with not having a special path for `akotlar/`.

Does the last commit address the goal?

cseed commented on monitoring/router.nginx.conf:
```diff
@@ -1,20 +1,45 @@
     location /monitoring/grafana/ {
         proxy_pass http://grafana/;
-        proxy_set_header Host internal.hail.is;
+        proxy_set_header Host $host;
```
I'm not sure why you changed this, it might be part of the problems you're seeing.  If memory serves, these services use Host in various redirects, so it needs to be host that originally came from the user.  That's `internal.hail.is`.

akotlar replied:
Grafana doesn't actually seem to care. I tried both internal.hail.is in monitoring.yaml, and grafana.hail.is, both worked. So either it isn't being used, or both get routed correctly, when host is $host.

cseed commented on monitoring/router.nginx.conf:
```diff
@@ -1,20 +1,45 @@
     location /monitoring/prometheus/ {
-        proxy_pass http://prometheus;
-        proxy_set_header Host internal.hail.is;
+        proxy_pass http://prometheus/;
```
You changed this to a trailing slash.  That means the location block now strips /monitor/prometheus which is probably why you're having trouble with prometheus.

akotlar replied:
> Let me know if you want me to take this over.

I have a professional interest in seeing this through.

akotlar replied:
> why you're having trouble with prometheus.

The behavior is totally unchanged from the configuration that had one monolithic server block. To figure out whether I had broken something, I restored John's config. Nothing changed, although it's possible I was seeing the effect of a crashloopbackoff say (though I checked that the pods were up each test, may have made a mistake).

The slash was added because the behavior looked inconsistent with Grafana's config. I read now that Prometheus expects the entire external url suffix, so what you say is consistent. We can add  "-web.route-prefix=/" to allow / to work.


> That means the location block now strips /monitor/prometheus

I thought so as well, but I wasn't seeing that behavior in logs. When I look at request logs of the grafana pod that received those requests, the request was /monitoring/grafana/ not /. That confused me, wasn't able to resolve last night.

I added the slash because 

I'll try again, and report the outcome.

akotlar replied:
> The behavior is totally unchanged from the configuration that had one monolithic server block. 

 I found a clue that may have contributed. I just tried modifying the monitoring.yaml config to add a prometheus feature. Did it, had a typo, sent to crashloopbackoff. Updated the correct setting, deployed again, and the pod wasn't sent back to containercreating. Restored the original monitoring.yaml config, deployed again, again no change. I needed to manually kill the pod for it to pick up the config.

cseed commented on monitoring/router.nginx.conf:
```diff
@@ -1,20 +1,45 @@
+        proxy_set_header Host $host;
+        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
+        proxy_set_header X-Forwarded-Host $host;
+        proxy_set_header X-Forwarded-Proto https;
```
This should have been set upstream by gateway based on the original request.  You probably don't want to overwrite them.  $host is now probably grafana.internal, which is only used for internal routing.

akotlar commented on monitoring/router.nginx.conf:
```diff
@@ -1,20 +1,45 @@
-    server_name _;
+    server_name grafana.*;
 
     location /monitoring/grafana/ {
```
There is something else that appears to be happening. 

Here is an experiment  that I ran last night:


```nginx
location /monitoring/grafana {
        proxy_pass http://grafana/;
        proxy_set_header Host $host;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Host $host;
        proxy_set_header X-Forwarded-Proto https;
    }

```

Result: Nothing goes to the monitoring namespace router. Something upstream of the router (or I guess in the proxy world, downstream, in any case, before anything is sent to the router) goes wrong.

Here is a picture. Where you see blank lines at the bottom, I simply added a few newlines to the router log stdout, to be able to observe routing events. With the above configuration, nothing gets routed to the router, or at least nothing is logged, and the url gets immediately changed to grafana.internal. It seems like a bug.

<img width="1620" alt="Screen Shot 2019-10-10 at 2 16 11 AM" src="https://user-images.githubusercontent.com/5543229/66577438-628b1c00-eb47-11e9-9fe2-9dacf46bab53.png">

akotlar commented:
Here is an example of the trailing slash issue:

What kibana sees:

"GET /monitoring/kibana/ui/fonts/inter_ui/Inter-UI-Bold.woff2 HTTP/1.1" 200 94840 "https://internal.hail.is/monitoring/kibana/app/kibana"

config:

```
location /monitoring/kibana/ {
        proxy_pass http://kibana/;
    }
```

It may not be inconsistent with trailing / on proxy_pass stripping the url, but it sure is confusing.

akotlar commented:
As an example of this slash issue, the following config (deployed right now) doesn't work.

```
location /monitoring/grafana {
        proxy_pass http://grafana/;
    }

location /monitoring/grafana/ {
        proxy_pass http://grafana/;
    }
```

Routing to https://internal.hail.is/monitoring/grafana appears to not hit the router (`k logs router-759c675b98-8mp67 -n monitoring -f`). Suggests problem is upstream of router-759.

akotlar commented:
What is the best way to test gateway?

cseed commented:
> The new config addresses the requested changes on the config.

I don't know what this means.  You didn't answer the question.

cseed commented:
> What is the best way to test gateway?

Deploy the new gateway.  Yes, it's live.  Yes, it might break stuff.  Good luck.

akotlar commented:
Regex isn't used to detect the optional slash because doing so doesn't really save any lines of code, because nginx does not directly allow proxy_pass with a trailing slash inside of regex-containing locations, without a rewrite rule. 

https://serverfault.com/questions/649151/nginx-location-regex-doesnt-work-with-proxy-pass

akotlar commented:
I also now suspect that the strange behavior I was seeing was due to caching on Chrome's side. This would explain why I was seeing nothing in server logs, and why behavior was inconsistent between browsers. I even watched logs of all 6 gateways (3 gateway, 3 internal), and the monitoring router, nothing.

Lastly, the proxy_set_header Host does not appear to be needed, so I have excluded it (tested with the Cluster dashboard). This also reduces the number of places we need to specify which external domain Grafana behind.

akotlar commented:
bump

akotlar commented:
bump

akotlar commented:
Assigned John, I think @cseed is busy

akotlar commented:
> I still generally find nginx and rules about trailing slashes confusing, but as far as I can tell this seems fine

What do you find confusing?

johnc1231 commented:
Yes, that helps a lot. Thanks :+1:
